lib/rouge/lexers/swift.rb

```diff
@@ -110,6 +110,8 @@ class deinit enum convenience extension final func import init internal lazy let
           groups Keyword, Text, Name::Variable
         end
 
+        rule %r/\\\.(#{id})/, Keyword::Type
```

Comment:
Is this valid Swift? From [the documentation](https://docs.swift.org/swift-book/ReferenceManual/Expressions.html#grammar_key-path-expression) looks like the syntax requires the type to occur immediately after the backslash?

Reply:
@pyrmont I think you're right, as far as I know the backslash has to be followed by a type name or `.` meaning 'infer the type from context'.  Could be wrong though, need an example.

Further, this lexer rule doesn't ever seem to fire because the backslash is eaten by the general Punctuation rule before we ever get here.

If @AlanQuatermain doesn't mind we could go ahead with the rest of this to add the missing keyword [which led me back to this PR, I thought it had been merged!]?

Reply:
@johnfairh I've been waiting for an answer to the above question. Is your suggestion to remove this line and merge?

Reply:
Yes, that's right.

Reply:
The keyword is optional, based on context. If the compiler already knows the type, then `\.isEmpty` is fine as an expression; otherwise you specify the type as part of the expression, i.e. `\String.isEmpty`.

For instance, this requires the type as part of the value expression:

```swift
var keyPath = \String.isEmpty   // Derived type is KeyPath<String, Bool>
```

This, however, does not, as the type is provided explicitly:

```swift
var keyPath: KeyPath<String, Bool> = \.isEmpty
```

The latter form is often used in functions, where the type is already specified by the function being called, either explicitly or implicitly via generic type resolution:

```swift
// This works because the argument explicitly says the root's type is `EnvironmentValues`.
public func environment<V>(_ keyPath: WritableKeyPath<EnvironmentValues, V>, _ value: V) -> some View
myView.environment(\.managedObjectContext, context)

// This works because the compiler, when resolving the types for the input to the function, 
// knows that whatever type it resolves for the `inObject` argument is also the root type of 
// the `KeyPath`.
public func readKeyPath<Item, Value>(_ keyPath: KeyPath<Item, Value>, from inObject: Item) -> Value
var empty = readKeyPath(\.isEmpty, "This is a String") // Item = String, Value = Bool, keyPath = KeyPath<String, Bool>
```

Reply:
Thanks for the detailed reply, @AlanQuatermain! :) I looked again at the documentation and either they changed it since I last checked (unlikely) or I simply lost the ability to read (much more likely) because it does indeed say that the type is optional. Nevertheless, your explanation was helpful!

Understanding better how the syntax works, I did wonder about the best approach. @johnfairh pointed out that the rule is never actually matched because of the higher precedence of the rule matching `%r{[()\[\]{}:;,?\\]}`. In the second-most recent push, I moved your rule higher up and added an example to the visual sample to make the comparison clearer.

The result looks to me a little strange. If the keypath includes the type, the slash is highlighted as punctuation, the type is highlighted as a type and the keypath-components (to use the terminology from Apple's docs) are highlighted as if they were functions. However, if the keypath doesn't include the type, the slash and the initial keypath-component is highlighted as a type.

The grammar in Apple's docs makes clear that a keypath can be quite complex. If Rouge is to highlight keypaths as distinct elements, I think we need a more comprehensive set of rules. Since your PR includes other improvements, can I suggest we follow @johnfairh's suggestion and remove this rule for now? A more comprehensive rule for keypaths can then be submitted as a separate PR. The most recent push implements that suggestion.

Let me know your thoughts!

Reply:
Any thoughts on the above @AlanQuatermain?

Reply:
I think the relative rarity of key path literals ought to mean we can put a more generic matcher for it at a higher precedence, because in 99.9% of cases it'll fail on the lack of backslash character—i.e. there aren't many cases where you'll begin a potential match with a backslash.

Either way, though, I don't especially mind. I'm primarily interested in having 'some' highlighted correctly, as my publisher is using Rouge in their publication flow and I want that keyword to show up properly in my book 😉

Reply:
@AlanQuatermain Apologies for perhaps being a bit dense but when you say a 'more generic matcher' are you referring to the rule that I removed in the most recent commit? Would you like that reverted? Your logic being that the fact that keypaths are highlighted differently depending on the presence of the type being not something to be worried about given their rarity?

Reply:
I'll poke my nose in here again because it's been quiet to suggest the PR be merged as-is for the next release to resolve the 'missing `some`' problem.

We can have a new issue if there is an example of poor key path formatting to work from: I agree with @pyrmont that the results of moving the proposed key path rule up are strange (I might use a stronger word!) and not something we should merge right now.

Reply:
Agree with @johnfairh. I've merged the PR as-is. Please feel free to open a new PR to improve key-path expression formatting.